backends/dpdk/dpdkArch.cpp

```diff
@@ -2374,6 +2402,191 @@ const IR::Node* SplitP4TableCommon::postorder(IR::SwitchStatement* statement) {
     return statement;
 }
 
+/* For regular tables, the direct counter array size is same as table size
+ * For learner tables, the direct counter/meter array size is 4 times the table size */
+int CollectDirectCounterMeter::getTableSize(const IR::P4Table * tbl) {
+    int tableSize = dpdk_default_table_size;
+    auto size = tbl->getSizeProperty();
+    if (size)
+        tableSize = size->asUnsigned();
+    return tableSize;
+}
+
+bool CollectDirectCounterMeter::checkMethodCallInAction(
+                const P4::ExternMethod *a, cstring method,
+                cstring &oneInstance, cstring instancename) {
+    bool found = false;
+    cstring externName = a->originalExternType->getName().name;
+    if (externName == "DirectMeter" || externName == "DirectCounter") {
+        auto di = a->object->to<IR::Declaration_Instance>();
+        cstring instanceName = di->name.name;
+        if (a->method->getName().name == method) {
+            if (instancename != "" && instanceName == instancename) {
+                found = true;
+            }
+            if (oneInstance == "")
+            oneInstance = instanceName;
+            // error if more than one count method found with different instance name
+            if (oneInstance != instanceName) {
+                ::error(ErrorType::ERR_UNEXPECTED, "%1% method for different %2% "
+                         "instances (%3% and %4%) called within same action",
+                         method, externName, oneInstance, instanceName);
+                return false;
+            }
+        }
+    }
+    return found;
+}
+
+bool CollectDirectCounterMeter::ifMethodFound(const IR::P4Action *a,
+                        cstring method, cstring instancename) {
+    cstring oneInstance = "";
+    bool found = false;
+    for (auto stmt : a->body->components) {
```

Comment:
you could do this by using a small visitor. it would be more robust and handle nested blocks, etc.

Reply:
I didn't get this comment correctly.
A visitor for BlockStatement and then calling visit(a->body) here?
Visitors for MethodCallStatement and AssignmentStatement  would also be needed I guess.

Reply:
You should probably create a new visitor instead of a loop